### DISCUSSION TITLE
[SPELLS] why they must be stored and maintained in database (almost all of them)

### DISCUSSION THREAD

**Discussion by Yehonal** [azerothcore/azerothcore-wotlk#5307] [Guides & Tips]
I have been reading many discussions about DBCs and spells with a lot of misunderstanding here and there.

So I would like to open this topic by trying to collect all the information we gained so far about the spell mechanism

1. **DBC Is not a single source of truth**: Blizzard has most of the spell mechanics written inside its own database, a small portion of them is extracted and converted to DBC to act as a way to instruct the client how to behave when a spell is cast (to avoid wasting network bandwidth with information that can be stored in the client instead, improving the latency as well). 
They extracted DBC files once per each client patch but, between a patch and another, the server-side spell database is possibly updated/extended with different values that are not needed by the client. Moreover, during the DBC extraction, there could be an alteration of the original values to fit the needs of the client (but not the ones of the server). Raw DBC files were initially used as a single source of truth by emulators to speed up the process of following the client progression, but using it as an immutable database for the server is totally wrong and that's why we need a table to override it (and coding them in some cases).

2. **Blizzard uses spells for almost everything:** Checking the [sniffs and the logs](https://wowgaming.github.io/wiki-en/classic-home.html) from the Classic you can notice that almost all the "effect" of a script are handled by a spell (when you have to set a unit flag, reducing the life, fake death status and even selecting creatures around etc.). Back in time we used to place almost all of this logic inside the code because we had very few information about the spell_dbc structure and any tables to operate with. But this is definitely an antipattern.  The only evidence of scripted spells is the ones with the "SPELL_AURA_DUMMY" attribute. But even in that case, we don't know if they are scripted with a database SAI-like system or by using pure code.

This topic could have some lacks of information, so feel free to contribute here with your comments. However, I hope that it can contribute to avoiding continuing to hardcode spells IDs within the core.

Moreover, you can extend this logic to basically everything: the code is not the place where you have to explicitly define such information (spells, areas, items etc.).

Thanks to [this system](https://github.com/azerothcore/azerothcore-wotlk/pull/3002) implemented into AzewrothCore based on [this idea](https://github.com/azerothcore/azerothcore-wotlk/issues/584#issue-253085904), now you are not bound anymore to DBC files and we can move this hardcoded stuff to the database.

**UPDATE:**

Few resources to explore the spell_dbc

* ACore dbc reader: https://github.com/wowgaming/node-dbc-reader
* DBC SQL version compatible with our structure: https://github.com/wowgaming/client-data/releases/tag/dbc_sql_v1
* Viewers and editors: https://wowgaming.github.io/wiki-en/client-dev-tools.html
* Spell DBC documentation: https://wowdev.wiki/DB/Spell#3.3.5.12340

**Comment by Kitzunu** [comment]
What we need is someone to clarify how spell_dbc works, because atm there is no way I am going to tinker with 230+ columns to try and get a spell to work when I can more easily do it as a spellscript.

(Which is also why I have disabled the feature on SpellWork, because I am not going to update, from TC's 96 columns in spell_dbc, to our 230+ with different names)

**Reply by Kitzunu** [reply]
as per this comment https://github.com/azerothcore/azerothcore-wotlk/pull/5298#issuecomment-820804144 it goes hand in hand what I just said. Two or three people know how spell_dbc works and how to apply fixes. It is hard to understand, hard to read instead of following a core structure.

Also spell_dbc uses names that we don't use anywhere in the entire core, for example AttributesExA-G, in the core and in the tools these are called AttributesEx2-7. Now it is not hard to understand that but why wasn't the names kept has we have used them?

**Reply by Yehonal** [reply]
The difference in names comes from the extractor @Winfidonarleyan used to convert the DBC to SQL, we can rename, improve and document that table at any time. Btw, the documentation about the table should be here: https://wowdev.wiki/DB/Spell

However, regarding that PR:

```C++
            case 20253:
                spellInfo->DmgClass = SPELL_DAMAGE_CLASS_MELEE;
                spellInfo->Attributes = SPELL_ATTR0_IMPOSSIBLE_DODGE_PARRY_BLOCK;
                spellInfo->SpellLevel = 0;
```

I don't see where's the issue in writing the following query to do the same thing:

```SQL
UPDATE spell_dbc SET DefenseType=2, Attributes=8192,SpellLevel=0 WHERE ID=20253;
```

However, we need also to import the full spell specification of those spells into that table before update it with new values.

**Reply by The-GhostRider** [reply]
@Kitzunu 
Start using **"WoW Spell Editor"** and quit **"SpellWork"**.
**"WoW Spell Editor" is the evolution** of "SpellWork", it not only allows you **to read all the spell data**, but also **allows you to edit them**.
**In addition it has an intuitive and easy to use interface**, if you need clarification **you can ask me what you want.**

**Reply by Yehonal** [reply]
TrinityCore has fewer fields because they still use DBC raw files as their main resource and they added some columns to override certain values they think must be overridden

**Reply by Yehonal** [reply]
I've added some link to the original post

**Reply by Kitzunu** [reply]
I can test wse, but from the screenshots I have seen I do prefer sw's view of displaying info.
But maybe it can be used to edit, which sw can't

**Reply by Kitzunu** [reply]
Could you add a link to where I can find it?

**Reply by The-GhostRider** [reply]
https://github.com/wowgaming/wiki-en/blob/master/wiki/client-dev-tools.md

**Reply by Yehonal** [reply]
I've added this link to the main post as well

**Reply by Kitzunu** [reply]
I will take a look at it when I get time, thanks Stiffler

**Reply by Yehonal** [reply]
> I can test wse, but from the screenshots I have seen I do prefer sw's view of displaying info.
> But maybe it can be used to edit, which sw can't

It's exactly what I do usually, SpellWork to view and search, while spell editor to edit

**Reply by kik0desu** [reply]
I agree with @Kitzunu , if they could rename the columns of ``spell_dbc`` the job would be a lot easier for everyone, especially people like me who are just starting out.

**Reply by Yehonal** [reply]
What's the current naming convention that we should follow in your opinion instead? and why? we could change the name of the properties in the code instead of the database ones.

**Reply by Kitzunu** [reply]
Me, I am just looking at everything that is numbered as ABC... I would prefer it to be 123.

Because to me SPELL_ATTR1_SOMETHING makes more sense than SPELL_ATTRA_SOMETHING, you see what I mean... Confusing. So that is why I think these things should be updated in DB and not the core

**Reply by kik0desu** [reply]
for example, `` DmgClass`` = `` DefenseType `` is not intuitive at all in my opinion. Please correct me if I am wrong.

**Reply by Yehonal** [reply]
I agree. After this change, we should provide an updated SQL dump here: https://github.com/wowgaming/client-data/releases/tag/dbc_sql_v1

About DmgClass, it's called also DefenseType around the web. Not sure why exactly. But I'm ok in changing it aswell. 
I think the most important thing is trying to keep the compatibility with external tools and also sync the name between db and code

**Reply by T1ti** [reply]
An issue with spell_dbc table in the DB is that it's specific to each emulator, they all have different structure and names.
While modifying the DBC will work with any emulator

This makes it harder to develop tools such as Spell Editor to use databases, because you would need to make one version of it for each emulator(this is the reason stoneharry hasn't implemented DB features to his tool)

A common effort between the emulators to standardize their column names to make it easier to develop tools would be required

**Reply by T1ti** [reply]
By the way WDBX should be avoided for spell.dbc usage, I found this issue while using it
https://github.com/WowDevTools/WDBXEditor/issues/103

**Reply by Yehonal** [reply]
I think that if Keira3 implemented a page to view and edit the spell_dbc table we wouldn't need any external tool to work with a spell in our DB. Keira3 could just read the [pre-filled spell_dbc table](https://github.com/azerothcore/azerothcore-wotlk/issues/5316), thus you can generate any kind of query you need to change/fix/add your spell.

 @FrancescoBorzi , @Helias do you think is it something that can be added in the keira3 roadmap?

**Reply by Footman** [reply]
what we need is new web tool to work with this. like you said to extend keira3 would be good approach.

**Reply by FrancescoBorzi** [reply]
yes, that's a good addition for Keira3. I will work on it, but my time is limited and I have plenty of other tasks to work on, so don't expect it too quickly.

**Comment by Yehonal** [comment]
@Kitzunu , @Stifler82 , @ErnestoGonzalezC 

I've created a tool to extract SQL from DBCs. It allows you to export only the rows you need. It's maintained by us so we can properly set the schema needed for the AzerothCore database structure. 

Check it out please: https://github.com/wowgaming/node-dbc-reader

Ideally, when we create a PR, we want to:

1. extract the query using that tool and set that query on top of our PR sql
2. creating the UPDATE query manually

Does it make sense for you?
